Append dependency's root path to CMAKE_PREFIX_PATH

osechet opened the pull request:
I made a simple modification to fix the issue #291. For each dependency, the `CONAN_<dep>_ROOT` is appended to `CMAKE_PREFIX_PATH`.

I tested it on Linux with a modified Protobuf package (https://github.com/osechet/conan-protobuf) where the test's CMakeLists calls `find_package()` and `protobuf_generate_cpp()`.

memsharded commented:
Thanks for your contribution! Just a question. We have tried to separate the generated ``conanbuildinfo.cmake`` in 2 parts. The first one just declares ``CONAN_XXX`` variables, but does not interfere at all with any other CMake variable or aspect. This is on purpose, so the ``conanbuildinfo.cmake`` can be included without any side effect at all, and let the consumer to use the declared variables as desired, if the available macros/functions doesn't fullfill they needs. Does this makes sense?

Do you think that we could do it inside the ``cmake_basic_setup()`` macro instead? Just after of before the setting of ``CMAKE_MODULE_PATH``, or just a two line macro ``conan_set_find_paths``.

I also think this is a good chance to try more exhaustively generator packages. So we can put the cmake (lets call the new one "cmakedev") inside a package, and start using it right now instead of having to wait until the next conan release. Would you like me to create it?

Thanks very much again.

osechet commented:
I'm not sure I really see the difference, but you know the project better than me :).
What if I use the `CONAN_CMAKE_MODULE_PATH` variable to set the `CMAKE_PREFIX_PATH` in the `conan_basic_setup()` macro (as it is done for `CMAKE_MODULE_PATH`)? I can also create a `CONAN_CMAKE_PREFIX_PATH` variable but its content would be exactly the same as the module one. So I don't think it's useful.

memsharded commented:
The difference is very subtle, and there is no difference in the final result, but more of a design issue.

There are conan users that don't want any interference at all with their CMake environment, variables, compilers, etc. But they want to know the include directories, library names, etc., of the installed conan packages. So they can ``include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)`` and be certain that it is not messing at all with their environment, and their variables. They might have already set their CMAKE_PREFIX_PATH to something and don't want conan to change that. They use the CONAN_XXX variables and do what they want to do with them, but they are not using the ``conan_basic_setup()`` helper function at all.

So, even if the final result could be the same if using ``conan_basic_setup()`` it is a two step process for those users that don't want it. First step is define CONAN_XXX variables with the desired values. Second step, in macros or functions, to apply those variables to actual CMAKE_XXX variables or other commands, that actually make CMake behave as expected.

I have reviewed the code, and I would say that in fact, yes, the content of the CONAN_CMAKE_MODULE_PATH would be exactly the same as the values that would be appended to CMAKE_PREFIX_PATH, have you cross checked it? Thanks!

lasote commented:
I agree, the assignment to CMAKE_PREFIX_PATH should be done in the setup method.

osechet commented:
As suggested, I now set the `CMAKE_PREFIX_PATH` in the `conan_basic_setup()` macro. I added a new internal macro (`conan_set_find_paths()`) to set the module path and the prefix path. Both of them are updated using the `CONAN_CMAKE_MODULE_PATH`. I also added a unit test to check if the generated content of the info file is the expected one.

memsharded commented:
Perfect, with unit test included, great job, thanks very much!

osechet commented:
Since this PR has been merged, issue #291 should probably be marked as fixed.

memsharded commented:
Yes, thanks. We have recently changed the management of issues, cause it was a bit confusing for users, so instead of closing it:
- Label as fixed
- Attach to milestone (next release 0.10)
- Close when released 0.10
